File: typescript/packages/node/src/commands/local_audio/ram/grep_ogg.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  IOResult,
  ResourceName,
  command,
  materialize,
  specOf,
  type ByteSource,
  type CommandFnResult,
  type CommandOpts,
  type PathSpec,
  type RAMAccessor,
} from '@struktoai/mirage-core'
import { transcribe } from '../utils.ts'
import { readRam } from './read.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder()

async function grepOggCommand(
  accessor: RAMAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length === 0 || texts.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('grep: missing operand\n') })]
  }
  const first = paths[0]
  if (first === undefined) return [null, new IOResult()]
  const pattern = texts[0] ?? ''
  const ignoreCase = opts.flags.i === true
  try {
    const raw = readRam(accessor, first)
    const text = DEC.decode(await materialize(transcribe(raw)))
    const flags = ignoreCase ? 'i' : ''
    const re = new RegExp(pattern, flags)
    if (re.test(text)) {
      const out: ByteSource = ENC.encode(text)
      return [
        out,
        new IOResult({ reads: { [first.stripPrefix]: raw }, cache: [first.stripPrefix] }),
      ]
    }
    return [
      null,
      new IOResult({
        exitCode: 1,
        reads: { [first.stripPrefix]: raw },
        cache: [first.stripPrefix],
      }),
    ]
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    return [
      null,
      new IOResult({
        exitCode: 1,
        stderr: ENC.encode(`grep: ${first.original}: failed to read as ogg: ${msg}\n`),
      }),
    ]
  }
}

export const RAM_LOCAL_AUDIO_GREP_OGG = command({
  name: 'grep',
  resource: ResourceName.RAM,
  spec: specOf('grep'),
  filetype: '.ogg',
  fn: grepOggCommand,
})

File: python/mirage/commands/local_audio/disk/grep_ogg.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import re
from collections.abc import AsyncIterator

from mirage.cache.index import IndexCacheStore
from mirage.commands.local_audio.utils import transcribe
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.disk.glob import resolve_glob
from mirage.core.disk.read import read_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("grep", resource="disk", spec=SPECS["grep"], filetype=".ogg")
async def grep_ogg(
    accessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    r: bool = False,
    R: bool = False,
    i: bool = False,
    v: bool = False,
    n: bool = False,
    c: bool = False,
    args_l: bool = False,
    w: bool = False,
    F: bool = False,
    E: bool = False,
    o: bool = False,
    m: str | None = None,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths or not texts:
        raise ValueError("grep: missing operand")
    paths = await resolve_glob(accessor, paths, index)
    try:
        pattern = texts[0]
        raw = await read_bytes(accessor, paths[0])
        chunks: list[bytes] = []
        async for chunk in transcribe(raw):
            chunks.append(chunk)
        full_text = b"".join(chunks).decode()
        flags = re.IGNORECASE if i else 0
        if re.search(pattern, full_text, flags):
            return full_text.encode(), IOResult(
                reads={paths[0].strip_prefix: raw},
                cache=[paths[0].strip_prefix])
        return None, IOResult(exit_code=1,
                              reads={paths[0].strip_prefix: raw},
                              cache=[paths[0].strip_prefix])
    except Exception as e:
        return None, IOResult(
            exit_code=1,
            stderr=f"grep: {paths[0].original}: failed to read as ogg: {e}".
            encode(),
        )
